subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepository.java

```diff
@@ -112,78 +135,128 @@ public String toString() {
             if (repositoryBlacklister.isBlacklisted(repositoryId)) {
-                result.failed(new ArtifactResolveException(component.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE));
+                result.failed(onBlacklisted.create());
                 return;
             }
 
-            try {
-                delegate.resolveArtifacts(component, result);
-            } catch (Throwable throwable) {
-                repositoryBlacklister.blacklistRepository(repositoryId, throwable);
-                result.failed(new ArtifactResolveException(component.getId(), throwable));
-            }
+            tryResolveAndMaybeBlacklist(result, operation, onError);
         }
 
-        @Override
-        public void resolveArtifact(ComponentArtifactMetadata artifact, ModuleSource moduleSource, BuildableArtifactResolveResult result) {
+        private <E extends Throwable, R extends ErroringResolveResult<E>> void performOperationWithRetries(R result,
+                                                                                                           Runnable operation,
+                                                                                                           Factory<E> onBlacklisted,
+                                                                                                           Transformer<E, Throwable> onError) {
             if (repositoryBlacklister.isBlacklisted(repositoryId)) {
-                result.failed(new ArtifactResolveException(artifact.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE));
+                result.failed(onBlacklisted.create());
                 return;
             }
 
-            try {
-                delegate.resolveArtifact(artifact, moduleSource, result);
-            } catch (Throwable throwable) {
-                repositoryBlacklister.blacklistRepository(repositoryId, throwable);
-                result.failed(new ArtifactResolveException(artifact.getId(), throwable));
+            tryResolveAndMaybeBlacklist(result, operation, onBlacklisted, onError);
+        }
+
+        private <E extends Throwable, R extends ErroringResolveResult<E>> void tryResolveAndMaybeBlacklist(R result,
+                                                                                                           Runnable operation,
+                                                                                                           Factory<E> onBlacklisted,
+                                                                                                           Transformer<E, Throwable> onError) {
+            tryResolveAndMaybeBlacklist(result, () -> {
+                operation.run();
+                return null;
+            }, onError);
+        }
+
+        private <E extends Throwable, R extends ErroringResolveResult<E>> void tryResolveAndMaybeBlacklist(R result, Callable<E> operation, Transformer<E, Throwable> onError) {
```

Comment:
We already have an ExponentialBackoff class in DefaultFileLockManager. We should extract that into a helper class.

Reply:
I saw it but it has been grown to something different than the initial intent. In particular there's no "file locking" involved here. I'd like to to keep things contained as much as possible.